### PR TITLE
Adds basic support for documentation

### PR DESCRIPTION
**Pull request by joeruello**
Hi there, hope PRs are welcome. I did this for my personal use but thought I would push it back upstream. This PR adds basic support for Documentation, specifically Sections and Decisions. 

It does not include images or documentation templates, but they should be pretty trivial to add.

I've done some basic testing (the entities make it into structurizr), I'll be using it myself over the next few days and will push any bug fixes.

### CODE REVIEW

**Comment by ChristianEder** [comment]
Hi Joe, this is awesome, thanks a lot for thinking of pushing this upstream, greatly appreciated!

Overall, the changes look good to me, I basically only have some minor findings in terms of how DTOs are handled, and also adding some documentation elements to the `test/api-compatibility.ts` test would be great. How would you like to handle those findings, I'm open to all approaches, e.g.:

- I can leave you some review comments
- I can make the changes directly in your PR branch (if you give permissions)

Just let me know and thanks again

**Comment by joeruello** [comment]
Either way is fine, I'm more than happy to make the changes my self.

I'm also working on adding support for `AutomaticLayout` and `PageSize` on views, once this PR is settled I'll apply same treatment to that branch and PR it as well.

**Comment by ChristianEder** [comment]
Great, then I'll leave you some comments - feel free to discuss / ask if some of my comments should'nt be clear.

File: src/core/documentation/decision.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
+import { Element } from "../model/element";
+import { Format } from "./format";
+import { DecisionStatus } from "./decisionStatus";
+
+export interface DecisionDto {
```
Comment (ChristianEder) [inline src/core/documentation/decision.ts]: I always also struggled with myself if I should add interfaces for the DTOs, in order to get some better support (e.g. auto completion) when writing toDto and fromDto methods. I finally settled with not adding the interfaces, since they will show up on the API exposed to the end user, and I dont want that.

What you could do, is to define, but not export the interface, so you'll at least have typing support within this single file

File: src/core/documentation/decision.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    id: string;
+    date: Date;
+    status: DecisionStatus;
+    element: Element | null;
+    title: string;
+    format: Format;
+    content: string;
+
+    constructor(
+        element: Element | null,
```
Comment (ChristianEder) [inline src/core/documentation/decision.ts]: The same reason as for the DTO interfaces basically goes for the constructors - the constructors are only to be called within this lib, not from the end user, which is why I kept them mostly parameterless in order not to expose too much detail on how they are constructed from DTOs.

Basically any other entity has an empty or nearly empty constructor, and a `fromDto` function. Like your `Documentation` class. Rule of thumb: if a class has a `toDto` function, it should also have a `fromDto`

File: src/core/documentation/documentation.ts [hunk 3]
```diff
@@ -0,0 +1,161 @@
+import { Model } from "../model/model";
+import { Element } from "../model/element";
+import * as assert from "assert";
+import { Section, SectionDto } from "./section";
+import { Decision, DecisionDto } from "./decision";
+import { SoftwareSystem } from "../model/softwareSystem";
+import { DecisionStatus } from "./decisionStatus";
+import { Format } from "..";
+
+interface DocumentationDto {
```
Comment (ChristianEder) [inline src/core/documentation/documentation.ts]: same remark as for the `DecisionDto` - please do not export Dto interfaces

File: src/core/documentation/documentation.ts [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+            title,
+            this.sections.size + 1,
+            format,
+            content
+        );
+        this.sections.add(section);
+        return section;
+    }
+
+    addDecision(
```
Comment (ChristianEder) [inline src/core/documentation/documentation.ts]: The rest of the code uses explicit `public` and `private` declarations

File: src/core/documentation/documentation.ts [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+            ),
+            decisions: Array.from(this.decisions.values()).map(decision =>
+                decision.toDto()
+            )
+        };
+    }
+
+    fromDto(dto: DocumentationDto) {
+        if (dto.sections) {
+            dto.sections.map(dto => {
```
Comment (ChristianEder) [inline src/core/documentation/documentation.ts]: this should be a `.forEach` instead of `.map` since you do not use the returned array. Same goes for the decisions below

File: src/core/documentation/documentation.ts [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+                const section = new Section(
```
Comment (ChristianEder) [inline src/core/documentation/documentation.ts]: as already mentioned above, please match this to the pattern you can see in https://github.com/ChristianEder/structurizr-typescript/blob/master/src/core/model/softwareSystem.ts#L31 :

- Create the sub entity using an empty constructor
- Call `.fromDto` on the sub entity

File: src/core/documentation/documentation.ts [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+    }
+
+    hydrate() {
+        this.sections.forEach(section => this.findElementAndHydrate(section));
+        this.decisions.forEach(decision =>
+            this.findElementAndHydrate(decision)
+        );
+    }
+
+    private findElementAndHydrate(section: {
```
Comment (ChristianEder) [inline src/core/documentation/documentation.ts]: nit: section isn't a good name here, since you use it for decisions as well. Perhaps `documentationElement`?

File: src/core/documentation/format.ts [hunk 8]
```diff
@@ -0,0 +1,4 @@
+export enum Format {
+    Markdown = "Markdown",
```
Comment (ChristianEder) [inline src/core/documentation/format.ts]: Other enums in the codebase do not use explicitly assigned string values, I am not sure on the effect this has on serialization, do you know? Is the structurizr API expecting string or int values on the wire?

File: src/core/model/workspace.ts [hunk 9]
```diff
@@ -1,10 +1,11 @@
...
-import { ViewSet } from "../view/viewSet";
+import { Model } from './model';
+import { ViewSet } from '../view/viewSet';
+import { Documentation } from '../documentation/documentation';
 
 export abstract class AbstractWorkspace {
     public id!: number;
-    public name: string = "";
-    public description: string = "";
+    public name: string = '';
```
Comment (ChristianEder) [inline src/core/model/workspace.ts]: I know that many people prefer double over single quotes, but in order to keep consistency, please change back to double quotes. Same goes for some of the `import` statements in this PR.

I am totally open to change from double to single quotes, but only consistently everywhere - and then, not in this PR.

File: src/core/documentation/decision.ts (hunk 1 above)
Comment (joeruello) [reply src/core/documentation/decision.ts]: I think i'll be consistent with the rest of the library, I feel typing isnt really too nessesary here.

File: src/core/documentation/documentation.ts (hunk 5 above)
Comment (joeruello) [reply src/core/documentation/documentation.ts]: :+1:

File: src/core/documentation/format.ts (hunk 8 above)
Comment (joeruello) [reply src/core/documentation/format.ts]: Yeah, odd. Typescript will default to outputting enums as indices unless you explicitly assign a value. I was going off the OpenAPI spec at `https://structurizr.com/help/web-api` which has the field defined as :
```
        format:
          type: string
          description: The content format type.
          enum:
            - Markdown
            - AsciiDoc
```

`InteractionStyle` is defined in a similar way, but apparently integers work just as well. My gut feeling is to follow the spec. Thoughts?

File: src/core/model/workspace.ts (hunk 9 above)
Comment (joeruello) [reply src/core/model/workspace.ts]: Rogue IDE formatting 🤦‍♂

File: src/core/index.ts [hunk 10]
```diff
@@ -1,29 +1,35 @@
...
-export * from './model/model';
-export * from './model/modelItem';
-export * from './model/person';
-export * from './model/relationship';
-export * from './model/relationships';
-export * from './model/softwareSystem';
-export * from './model/staticStructureElement';
-export * from './model/tags';
-export * from './model/workspace';
+export * from "./model/container";
```
Comment (joeruello) [inline src/core/index.ts]: This file was all single-quoted, I converted to double quotes for consistency as I was adding some exports. Let me know if not in scope and I'll revert.

File: src/core/documentation/documentation.ts [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+        element?: Element;
+    }) {
+        if (documentationElement.elementId) {
+            documentationElement.element = this.model.getElement(
+                documentationElement.elementId
+            );
+        }
+    }
+
+    private checkSectionIsUnqiue(title: string, element?: Element) {
```
Comment (joeruello) [inline src/core/documentation/documentation.ts]: The previous way I implemented these checks was trying to be too clever and buggy, Instead, I fell back to a straight for of the java client's implementation.

File: src/core/documentation/documentation.ts [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+        if (!element) {
+            this.sections.forEach(section => {
+                if (!section.element && section.title === title) {
+                    throw new Error(
+                        `A section with a title of ${title} lready exists for this workspace.`
```
Comment (ChristianEder) [inline src/core/documentation/documentation.ts]: nit: typo 'lready'

Comment (ChristianEder) [reply src/core/documentation/documentation.ts]: same also below

File: src/core/documentation/format.ts (hunk 8 above)
Comment (ChristianEder) [reply src/core/documentation/format.ts]: I agree with you - following the spec is the better approach here :-) This just means a reminder to myself that I'll have to check other enums as well.

File: src/core/index.ts (hunk 10 above)
Comment (ChristianEder) [reply src/core/index.ts]: This is fine, thanks!

**Comment by ChristianEder** [comment]
Besides the 1 open comment regarding the `lready` typo, this looks great to me. I'll give it a test run tomorrow and then merge it in, thanks again!

Do you have a twitter handle I should add when announcing the new version?

**Comment by joeruello** [comment]
Not that I've used since 2012 😅. I added also just opened another PR for `AutomaticLayout`, guessing it'll make sense to release them together.

File: src/core/documentation/documentation.ts [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+    private model: Model;
+    public sections: Set<Section> = new Set();
+    public decisions: Set<Decision> = new Set();
+
+    constructor(model: Model) {
+        this.model = model;
+    }
+
+    public addSection(
+        element: Element,
```
Comment (ChristianEder) [inline src/core/documentation/documentation.ts]: the element parameter is optional - using `element?: Element` won't compile though without moving the parameter to the end. I'd suppose to use `element: Element | undefined`

File: src/core/documentation/documentation.ts [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+            order: this.sections.size + 1,
+            format,
+            content
+        });
+        this.sections.add(section);
+        return section;
+    }
+
+    public addDecision(
+        element: SoftwareSystem,
```
Comment (ChristianEder) [inline src/core/documentation/documentation.ts]: Same here: the softwareSystem should be optional (and better renamed): `softwareSystem: SoftwareSystem | undefined`.

Also: the parameter is not used in the code below, which I believe it should, right?

File: src/core/documentation/documentation.ts (hunk 13 above)
Comment (ChristianEder) [reply src/core/documentation/documentation.ts]: this will then also require changing the call to this function from the test workspace to use `undefined` instead of  `null`

**Comment by ChristianEder** [comment]
I just tried to get this running on my machine, and except for the findings I created the new comments for, it worked out.

**Comment by ChristianEder** [comment]
Hey Joe, is there any way I can support you in doing the final changes before merging this in?

**Comment by ChristianEder** [comment]
Again a little poke - if I can be of any help, let me know. If you want me to, I can also take over and finalize the changes, just let me know.
